src/python/gudhi/barycenter.py

```diff
@@ -0,0 +1,342 @@
+import ot
+import numpy as np
+import scipy.spatial.distance as sc
+
+
+# This file is part of the Gudhi Library - https://gudhi.inria.fr/ - which is released under MIT.
+# See file LICENSE or go to https://gudhi.inria.fr/licensing/ for full license details.
+# Author(s):       Theo Lacombe
+#
+# Copyright (C) 2019 Inria
+#
+# Modification(s):
+#   - YYYY/MM Author: Description of the modification
+
+
+def _proj_on_diag(w):
+    '''
+        Util function to project a point on the diag.
+    '''
+    return np.array([(w[0] + w[1])/2 , (w[0] + w[1])/2])
+
+
+def _proj_on_diag_array(X):
+    '''
+    :param X: (n x 2) array encoding the points of a persistent diagram.
+    :returns: (n x 2) array encoding the (respective orthogonal) projections of the points onto the diagonal
+    '''
+    Z = (X[:,0] + X[:,1]) / 2.
+    return np.array([Z , Z]).T
+
+
+def _build_dist_matrix(X, Y, p=2., q=2.):
+    '''
+    :param X: (n x 2) numpy.array encoding the (points of the) first diagram.
+    :param Y: (m x 2) numpy.array encoding the second diagram.
+    :param q: Ground metric (i.e. norm l_q).
+    :param p: exponent for the Wasserstein metric.
+    :returns: (n+1) x (m+1) np.array encoding the cost matrix C. 
+                For 1 <= i <= n, 1 <= j <= m, C[i,j] encodes the distance between X[i] and Y[j], while C[i, m+1] (resp. C[n+1, j]) encodes the distance (to the p) between X[i] (resp Y[j]) and its orthogonal proj onto the diagonal.
+                note also that C[n+1, m+1] = 0  (it costs nothing to move from the diagonal to the diagonal).
+                Note that for lagrangian_barycenter, one must use p=q=2. 
+    '''
+    Xdiag = _proj_on_diag_array(X)
+    Ydiag = _proj_on_diag_array(Y)
+    if np.isinf(q):
+        C = sc.cdist(X, Y, metric='chebyshev')**p
+        Cxd = np.linalg.norm(X - Xdiag, ord=q, axis=1)**p
+        Cdy = np.linalg.norm(Y - Ydiag, ord=q, axis=1)**p
+    else:
+        C = sc.cdist(X,Y, metric='minkowski', p=q)**p
+        Cxd = np.linalg.norm(X - Xdiag, ord=q, axis=1)**p
+        Cdy = np.linalg.norm(Y - Ydiag, ord=q, axis=1)**p
+    Cf = np.hstack((C, Cxd[:,None]))
+    Cdy = np.append(Cdy, 0)
+
+    Cf = np.vstack((Cf, Cdy[None,:]))
+
+    return Cf
+
+
+def _optimal_matching(X, Y):
```

Comment:
Is that more or less the function that Hari needs, and which could become public in the future?

Reply:
Well, I do not really know what Hari needs.
But yes, I agree that this function, or something similar, should be public in the future (in `representations.` I guess?)

Reply:
Once your other branch is merged, will you be able to remove this function completely?

Reply:
I think so.